mypy/checkexpr.py

```diff
@@ -1743,15 +1743,9 @@ def visit_func_expr(self, e: FuncExpr) -> Type:
         inferred_type = self.infer_lambda_type_using_context(e)
         if not inferred_type:
             # No useful type context.
-            ret_type = e.expr().accept(self.chk)
-            if not e.arguments:
-                # Form 'lambda: e'; just use the inferred return type.
-                return CallableType([], [], [], ret_type, self.named_type('builtins.function'))
-            else:
-                # TODO: Consider reporting an error. However, this is fine if
-                # we are just doing the first pass in contextual type
-                # inference.
-                return AnyType()
+            ret_type = self.accept(e.expr())
+            fallback = self.named_type('builtins.function')
+            return callable_type(e, fallback, ret_type)
```

Comment:
I don't know how does the shadowing work, but it seems to work (`lambda x: x` does not return the type of an outer `x`)